src/namespace.rs

```diff
@@ -0,0 +1,167 @@
+use std::convert::Into;
+use std::env::current_dir;
+use std::env::set_current_dir;
+use std::fs;
+use std::fs::File;
+use std::io;
+use std::io::BufRead;
+use std::io::BufReader;
+use std::os::fd::AsFd;
+use std::os::fd::AsRawFd;
+use std::os::unix::fs::MetadataExt;
+use std::path::PathBuf;
+use std::str::FromStr;
+
+use libc::setns;
+use libc::CLONE_NEWNS;
+
+use crate::Error;
+use crate::ErrorExt;
+use crate::Pid;
+
+pub(crate) struct NsInfo {
+    tgid: Pid,
+    nstgid: Pid,
+    need_setns: bool,
+    mntns_path: Option<PathBuf>,
+    oldns: File,
+    // From https://github.com/torvalds/linux/commit/b01c1f69c8660eaeab7d365cd570103c5c073a02, we see
+    // once finished we setns to old namespace, which also sets the current working directory (cwd) to "/",
+    // trashing the cwd we had. So adding the current working directory to be part of `NsInfo` and restoring
+    // it in the `Drop` call.
+    oldcwd: PathBuf,
+}
+
+fn get_nspid(pid: Pid) -> Result<(Pid, Pid), Error> {
+    let fname = format!("/proc/{pid}/status");
+    let file = File::open(&fname).context("faild to open `{fname}`")?;
+    let reader = BufReader::new(file);
+    let (mut tgid, mut nstgid) = (pid, pid);
+    let mut found = false;
+
+    for line in reader.lines() {
+        match line {
+            Ok(line) => {
+                /* Use tgid if CONFIG_PID_NS is not defined. */
+                if line.contains("Tgid:") {
+                    if let Some(num) = line
+                        .split("Tgid:")
+                        .last()
+                        .and_then(|s| s.split_whitespace().next_back())
+                    {
+                        let id = Pid::from_str(num).map_err(|e| {
+                            io::Error::new(io::ErrorKind::InvalidInput, format!("{e}"))
+                        })?;
+                        tgid = id;
+                        nstgid = id;
+                        found = true;
+                    }
+                }
+                if line.contains("NStgid:") {
+                    if let Some(num) = line
+                        .split("NStgid:")
+                        .last()
+                        .and_then(|s| s.split_whitespace().next_back())
+                    {
+                        nstgid = Pid::from_str(num).map_err(|e| {
+                            io::Error::new(io::ErrorKind::InvalidInput, format!("{e}"))
+                        })?;
+                        break;
+                    }
+                }
+            }
+            Err(e) => return Err(e.into()),
+        }
+    }
+
+    if !found {
+        unreachable!("{}", format!("failed to get Tgid/NStgid from {fname}"));
```

Comment:
As mentioned, let's not panic here on error but handle this case more gracefully.